File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        int curr = 0, prev = -1, n = nums.size();
        vector<vector<int>> dp(n+1, vector<int>(n+1, -1));
        return f(curr, prev, nums, dp);
    }
    int f(int curr, int prev, vector<int>&nums, vector<vector<int>>&dp){
        if(curr >= nums.size()) return 0;
        if(dp[curr][prev+1] != -1) return dp[curr][prev+1];
        int take = 0;
        if(prev == -1 || nums[curr] > nums[prev]) take = 1 + f(curr+1, curr, nums, dp);
        int notTake = 0 + f(curr+1, prev, nums, dp);
        dp[curr][prev+1] = max(take, notTake);
        return dp[curr][prev+1];
    }
};

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [1] * n
        for i in range(1, n):
            for j in range(0, i):
                if nums[i] > nums[j] and dp[i] <= dp[j]:
                    dp[i] = dp[j]+1
        return max(dp)